IntegrationTests/Sources/main.swift

```diff
@@ -1,16 +1,142 @@
 import Foundation
 import mParticle_Apple_SDK
 
+// Listener for tracking upload events
+class EventUploadWaiter: NSObject, MPListenerProtocol {
+    private var uploadCompletedSemaphore: DispatchSemaphore?
+    var mparticle = MParticle.sharedInstance()
+
+    @discardableResult
+    func wait(timeout: Int = 10) -> Bool {
+        mparticle.upload()
+        let semaphore = DispatchSemaphore(value: 0)
+        uploadCompletedSemaphore = semaphore
+        
+        let timeoutTime = DispatchTime.now() + .seconds(timeout)
+        let result = semaphore.wait(timeout: timeoutTime)
+        
+        uploadCompletedSemaphore = nil
+        
+        return result == .success
+    }
+    
+    func onNetworkRequestFinished(_ type: MPEndpoint, 
+                                  url: String, 
+                                  body: NSObject, 
+                                  responseCode: Int) {
+        if type == .events {
+            uploadCompletedSemaphore?.signal()
+        }
+    }
+    
+    func onNetworkRequestStarted(_ type: MPEndpoint, url: String, body: NSObject) {}
+}
+
+// Test 1: Simple Event
+func testSimpleEvent(mparticle: MParticle, uploadWaiter: EventUploadWaiter) {
+    mparticle.logEvent("Simple Event Name", eventType: .other, eventInfo: ["SimpleKey": "SimpleValue"])
+    uploadWaiter.wait()
+}
+
+// Test 2: Log Event with Custom Attributes and Custom Flags
+// Based on ViewController.m logEvent method (lines 131-147)
+func testEventWithCustomAttributesAndFlags(mparticle: MParticle, uploadWaiter: EventUploadWaiter) {
+    let event = MPEvent(name: "Event Name", type: .transaction)
+    
+    // Use static date instead of Date() for deterministic testing
+    let staticDate = Date(timeIntervalSince1970: 1700000000) // Fixed timestamp: 2023-11-14 22:13:20 UTC
+    
+    // Add custom attributes including string, number, date, and nested dictionary
+    event?.customAttributes = [
+        "A_String_Key": "A String Value",
+        "A Number Key": 42,
+        "A Date Key": staticDate,
+        "test Dictionary": [
+            "test1": "test",
+            "test2": 2,
+            "test3": staticDate
+        ]
+    ]
+    
+    // Custom flags - sent to mParticle but not forwarded to other providers
+    event?.addCustomFlag("Top Secret", withKey: "Not_forwarded_to_providers")
+    
+    // Log the event
+    if let event = event {
+        mparticle.logEvent(event)
+    }
+    uploadWaiter.wait()
+}
+
+// Test 3: Log Screen
+// Based on ViewController.m logScreen method (lines 149-151)
+func testLogScreen(mparticle: MParticle, uploadWaiter: EventUploadWaiter) {
+    mparticle.logScreen("Home Screen", eventInfo: nil)
+    uploadWaiter.wait()
+}
+
+// Test 4: Log Commerce Event with Product and Transaction
+// Based on ViewController.m logCommerceEvent method (lines 153-180)
```

Comment:
```suggestion
// Based on ViewController.m logCommerceEvent method
```

Line numbers are highly likely to change